moon/controller_round3.py

```diff
@@ -302,85 +308,88 @@ def on_artf(self, data):
+                                self.nasa_roll = math.atan2(sinr_cosp, cosr_cosp);
+
+                                sinp = 2 * (qw * qy - qz * qx);
+                                if abs(sinp) >= 1:
+                                    self.nasa_pitch = math.copysign(math.pi / 2, sinp);
+                                else:
+                                    self.nasa_pitch = math.asin(sinp);
+
+                                self.nasa_pitch = - self.nasa_pitch # for subsequent calculations, up is positive and down is negative
+
+                                siny_cosp = 2 * (qw * qz + qx * qy);
+                                cosy_cosp = 1 - 2 * (qy * qy + qz * qz);
+                                self.nasa_yaw = math.atan2(siny_cosp, cosy_cosp);
+                                print (self.sim_time, "app: True pose received: xyz=[%f,%f,%f], roll=%f, pitch=%f, yaw=%f" % (origin[0],origin[1],origin[2],self.nasa_roll, self.nasa_pitch, self.nasa_yaw))
+
+                                print(self.sim_time, "app: Final frame x=%d y=%d w=%d h=%d, nonblack=%d" % (data[1], data[2], data[3], data[4], data[5]))
+                                angle_x = math.atan( (CAMERA_WIDTH / 2 - (img_x + img_w/2) ) / float(CAMERA_FOCAL_LENGTH))
+                                angle_y = math.atan( (CAMERA_HEIGHT / 2 - (img_y + img_h/2) ) / float(CAMERA_FOCAL_LENGTH))
+
+                                distance = self.interpolate_distance((img_w + img_h) / 2)
+                                ax = self.nasa_yaw + angle_x
+                                ay = self.nasa_pitch + angle_y
+
+                                if self.use_gimbal:
+                                    # gimbal changes the actual angle dynamically so pitch needs to be offset
+                                    ay += min(math.pi / 4.0, max(-math.pi / 8.0, self.camera_angle - self.nasa_pitch))
+                                else:
+                                    ay += self.camera_angle
+
+
+                                x, y, z = self.nasa_xyz
+                                print("Using pose: xyz=[%f %f %f] orientation=[%f %f %f]" % (x, y, z, self.nasa_roll, self.nasa_pitch, self.nasa_yaw))
+                                print("In combination with view angle %f %f and distance %f" % (ax, ay, distance))
+                                ox = math.cos(ax) * math.cos(ay) * distance
+                                oy = math.sin(ax) * math.cos(ay) * distance
+                                oz = math.sin(ay) * distance
+                                self.cubesat_location = (x+ox, y+oy, z+oz)
+                                print (self.sim_time, "app: Object offset calculated at: [%f %f %f]" % (ox, oy, oz))
+                                print (self.sim_time, "app: Reporting estimated object location at: [%f,%f,%f]" % (x+ox, y+oy, z+oz))
+
+                                def process_apriori_object(response):
```

Comment:
callback in callback - is it really good idea??

Reply:
It's a common paradigm (at least in JavaScript) called 'Callback Hell', however there may certainly be a better way to achieve the same in Python.

Reply:
There is a reason for the name of this paradigm :wink: When the need for callback in a callback appears, it might be an indication that maybe a separate stack/thread/node is needed where the sequence of call backs can be handled as regular blocking calls.